usr/lib/common/pkey_utils.c

```diff
@@ -602,3 +702,343 @@ CK_RV pkey_aes_cmac(OBJECT *key_obj, CK_BYTE *message,
+        memcpy(param.P256.vp, pkey_attr->pValue + 32, 32);
+        *sig_len = 2 * 32;
+        break;
+    case curve_p384:
+        if (hash_len > 48)
+            hash_len = 48;
+        off = 48 - hash_len;
+        memcpy(param.P384.hash + off, hash, hash_len);
+        memcpy(param.P384.priv, pkey_attr->pValue, 48);
+        memcpy(param.P384.vp, pkey_attr->pValue + 48, 32);
+        *sig_len = 2 * 48;
+        break;
+    case curve_p521:
+        if (hash_len > 66)
+            hash_len = 66;
+        /* Note that the pkey for p521 has 80 + 32 bytes. */
+        off = 80 - hash_len;
+        memcpy(param.P521.hash + off, hash, hash_len);
+        memcpy(param.P521.priv, pkey_attr->pValue, pkey_attr->ulValueLen - 32);
+        memcpy(param.P521.vp, pkey_attr->pValue + pkey_attr->ulValueLen - 32, 32);
+        *sig_len = 2 * 66;
+        break;
+    default:
+        TRACE_ERROR("Could not determine the curve type.\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    if (sig == NULL) {
+        /* length_only requested, just return with *sig_len */
+        ret = CKR_OK;
+        goto done;
+    }
+
+    /* Set random number and function code */
+    switch (curve_type) {
+    case curve_p256:
+        fc = KDSA_ENCRYPTED_ECDSA_SIGN_P256 | 0x80;
+        if (rng_cb != NULL) {
+            rng_cb(param.P256.rand, sizeof(param.P256.rand));
+            fc = KDSA_ENCRYPTED_ECDSA_SIGN_P256;
+        }
+        break;
+    case curve_p384:
+        fc = KDSA_ENCRYPTED_ECDSA_SIGN_P384 | 0x80;
+        if (rng_cb != NULL) {
+            rng_cb(param.P384.rand, sizeof(param.P384.rand));
+            fc = KDSA_ENCRYPTED_ECDSA_SIGN_P384;
+        }
+        break;
+    case curve_p521:
+        fc = KDSA_ENCRYPTED_ECDSA_SIGN_P521 | 0x80;
+        if (rng_cb != NULL) {
+            rng_cb(param.P521.rand, sizeof(param.P521.rand));
+            fc = KDSA_ENCRYPTED_ECDSA_SIGN_P521;
+        }
+        break;
+    default:
+        /* Cannot occur */
+        goto done;
+    }
+
+retry:
+    /* Call CPACF */
+    rc = s390_kdsa(fc, param.buff, NULL, 0);
+    switch (rc) {
+    case 0:
+        ret = CKR_OK;
+        break;
+    case 1:
+        switch (curve_type) {
+        case curve_p256:
+            TRACE_ERROR("%s rc from KDSA = 1 and C = %02X.\n",
+                        __func__, param.P256.c);
+            break;
+        case curve_p384:
+            TRACE_ERROR("%s rc from KDSA = 1 and C = %02X.\n",
+                        __func__, param.P384.c);
+            break;
+        case curve_p521:
+            TRACE_ERROR("%s rc from KDSA = 1 and C = %02X.\n",
+                        __func__, param.P521.c);
+            break;
+        default:
+            break;
+        }
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+        break;
+    case 2:
+        TRACE_ERROR("%s rc from KDSA = 2\n", __func__);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+        break;
+    case 3:
```

Comment:
How can this happen?  Partial completion is already taken care of by `s390_kdsa`.

Reply:
OK, removed "case 3".